### PR TITLE
syntax: Simplify parsing of paths

### PR DESCRIPTION
**Pull request by petrochenkov**
Discern between `Path` and `Path<>` in AST (but not in HIR).
Give span to angle bracketed generic arguments (`::<'a, T>` in `path::segment::<'a, T>`).

This is a refactoring in preparation for https://internals.rust-lang.org/t/macro-path-uses-novel-syntax/5561/3, but it doesn't add anything to the grammar yet.

r? @jseyfried

### CODE REVIEW

File: src/test/ui/span/macro-ty-params.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    //~| ERROR generic arguments in macro path
+    foo::<>!();
+    //~^ ERROR generic arguments in macro path
+    //~| ERROR generic arguments in macro path
+    //~| ERROR generic arguments in macro path
+    m!(MyTrait<>);
+    //~^ ERROR generic arguments in macro path
+    //~| ERROR generic arguments in macro path
+    //~| ERROR generic arguments in macro path
+    //~| ERROR generic arguments in macro path
```
Comment (petrochenkov) [inline src/test/ui/span/macro-ty-params.rs]: I don't know why `resolve_macro_to_def` is called so many times on macro paths, but it seems to be an orthogonal issue (and maybe a performance concern).

Comment (petrochenkov) [inline src/test/ui/span/macro-ty-params.rs]: Ah, I see, it may be undetermined a few times and then resolved later. So the generic check need to be done only when resolution is successful.

Comment (jseyfried) [inline src/test/ui/span/macro-ty-params.rs]: It might make sense to do this check in `fn finalize_current_module_macro_resolutions` in `librustc_resolve/macros.rs`.

**Comment by jseyfried** [comment]
Did a first pass-through review, LGTM modulo [comment](https://github.com/rust-lang/rust/pull/43438#discussion_r128978976).

I'd like to do a second pass over the parser changes tomorrow, otherwise r=me.

**Comment by petrochenkov** [comment]
>I'd like to do a second pass over the parser changes tomorrow

The idea is that `parse_path_segments_without_colons`/`parse_path_segments_with_colons`/`parse_path_segments_without_types` are merged into a single `parse_path_segments` calling `parse_path_segment` in a loop, which in its turn does all the interesting job. Method calls and field accesses now also use this `parse_path_segment` instead of custom code.
This is mostly a rewrite, so reading the new code may be more convenient than the diff.

File: src/test/ui/span/macro-ty-params.rs (hunk 1 above)
Comment (petrochenkov) [inline src/test/ui/span/macro-ty-params.rs]: Looks like the necessary information is already lost in `finalize_current_module_macro_resolutions`, so I kept it in `resolve_macro_to_def`.

**Comment by jseyfried** [comment]
Looks great! r=me
